File: leetcode/two_pointers/slow_fast_pointers/MoveZeroes.java
/*
283. Move Zeroes
Easy
Topics
Companies
Hint
Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Note that you must do this in-place without making a copy of the array
*/

class MoveZeroes {
	/*
	Approach: 2 pointers (fast, slow)
	Time: O(n) Space:O(1)
	*/
    private static void moveZeroes(int[] nums) {
        int insertPos = 0;
        for(int i=0; i<nums.length; i++) {
        	if (nums[i] != 0) nums[insertPos++] = nums[i];
        }

        while(insertPos < nums.length) nums[insertPos++] = 0;
        return;
    }

    public static void main(String[] args) {
    	int[] nums = {0,1,0,3,12};
    	moveZeroes(nums);
    	for(int i: nums) System.out.print(i + " ");
    }
}

File: leetcode/two_pointers/logic/move_zeroes.py
"""
283. Move Zeroes
Solved
Easy
Topics
Companies
Hint
Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Note that you must do this in-place without making a copy of the array.
"""


class MoveZeroes:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if len(nums) <= 1:
            return

        # Index to be written at
        writePtr = 0

        # Index to be read at
        for readPtr in range(len(nums)):
            if nums[readPtr] != 0:
                nums[writePtr] = nums[readPtr]
                writePtr += 1

        while writePtr < len(nums):
            nums[writePtr] = 0
            writePtr += 1
